### ISSUE TITLE
Improvement: Preload Pages and Cache loaded pages

### ISSUE THREAD

**Issue by nikischin** [wojtekmaj/react-pdf#1816]
### Before you start - checklist

- [X] I understand that React-PDF does not aim to be a fully-fledged PDF viewer and is only a tool to make one
- [X] I have checked if this feature request is not already reported

### Description

I just integrated this library in my App and it’s a huge upgrade compared to an iframe from a visual perspective. However, talking about performance it cannot compete with an iframe. And that is totally ok, though I was wondering if it would be possible in a future version to automatically preload the next page and keep the previous few pages in memory maybe? Because changing pages really takes some time and even when you go back to pages previously rendered they need to load again (without even showing the loading indicator and just being white).

### Proposed solution

Automatically preload the next (or even next few) pages.
Keep previous few pages in Memory. Also preload the previous page if not in memory already.

Thank you!!

**Comment by nikischin** [comment]
@wojtekmaj how can a Feature Request get stale? :)

**Comment by wojtekmaj** [comment]
I guess you could make a proxy component like OffscreenPage by rendering Page component, with text and annotation layers turned off, and customRenderer noop function? 🤔 You also have useDocumentContext hook which enables you to programatically get pages as you need.

**Comment by nikischin** [comment]
So I guess this is nothing which could be added directly to this library then? :)

**Comment by vipcxj** [comment]
Any news? Just switching one page causes a huge delay, which makes the user experience very bad. If this library is not planning to fix this issue, is there any good solution for those of us using it? I'm not asking for the first load to be very fast, but at least the pages that have already been loaded should be able to load instantly again. Specifically, how to cache pages.
@nikischin Do you have any solutions?

**Comment by nikischin** [comment]
> Any news? Just switching one page causes a huge delay, which makes the user experience very bad. If this library is not planning to fix this issue, is there any good solution for those of us using it? I'm not asking for the first load to be very fast, but at least the pages that have already been loaded should be able to load instantly again. Specifically, how to cache pages. [@nikischin](https://github.com/nikischin) Do you have any solutions?

Hi @vipcxj unfortunately I do not have any solution but would be interested in one as well. I would prefer this to be implemented with the library and not with some custom implementation based on hooks.

**Comment by vipcxj** [comment]
@nikischin I asked AI, and it gave me several solutions. One is to cache the canvas yourself, but that failed and caused the PDF to display blank. Another is to render all pages at once, placing each page in a separate div and controlling its display with display:none/block. This allows for instant page turning, and I think this is the best solution. However, my PDF doesn't have many pages, so I'm not sure if a PDF with a very large number of pages would cause excessively long initial loading times or even fail to load altogether.

**Comment by nikischin** [comment]
> [@nikischin](https://github.com/nikischin) I asked AI, and it gave me several solutions. One is to cache the canvas yourself, but that failed and caused the PDF to display blank. Another is to render all pages at once, placing each page in a separate div and controlling its display with display:none/block. This allows for instant page turning, and I think this is the best solution. However, my PDF doesn't have many pages, so I'm not sure if a PDF with a very large number of pages would cause excessively long initial loading times or even fail to load altogether.

Yes @vipcxj , just as @wojtekmaj mentioned something like this should be possible. 

> I guess you could make a proxy component like OffscreenPage by rendering Page component, with text and annotation layers turned off, and customRenderer noop function? 🤔 You also have useDocumentContext hook which enables you to programatically get pages as you need.

I however find it quite challenging to maintain functionality myself which is rather deep in an external library (in this case react-pdf)
